### PR TITLE
Cody/airgrid efficiency

### PR DESCRIPTION
**Pull request by CWelshE**
Some basic efficiency suggestions that could be implemented in a few of the engine-specific files. :slightly_smiling_face:

Most of it is related to the three following factors:

- Make sure to combine operations in loops that might currently be separated.
- Many expensive operations can use memoization for constant-time efficiency.
- See if you don't need to perform an operation at all, and instead integrate it into the work that something else is already doing.

Let me know if this is useful!

### CODE REVIEW

File: src/engine/evaluate.ts [hunk 1]
```diff
@@ -7,26 +7,39 @@ import {
...
     })
   );
 
-  return rules.every((rule) => {
-    // TODO: allow other reducers...
-    //     // const reducer = reducers[rule.reducer.name](rule.reducer.args);
+  // Instead of iterating through everything, assume everything matches
+  // first, and exit the loop early if we find anything that doesn't.
+  // Further: Can the rules array be sorted in some way? If so, we can
+  // possibly use a binary search for O(log n) performance.
```
Comment (naripok) [inline src/engine/evaluate.ts:32]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every

`every` does short-circuit immediately on the finding of a falsy element.
Does this change provide performance improvements anyway?

File: src/engine/filters/math.ts [hunk 2]
```diff
@@ -1,7 +1,29 @@
-export const dotProduct = (vectorA: number[], vectorB: number[]): number => {
-  return vectorA.reduce((acc, cur, idx) => acc + cur * vectorB[idx], 0);
+// These are potentially expensive function calls. I would recommend
+// memoization here to reduce the total number of calls necessary after
+// the first one. For instance:
+//
+// dotProduct(1, 2) => (some result) is a required operation every time.
+// dotProduct(1, 2, memo[1, 2]) => (some result) can simply access the
+// `memo` array/variable/hash table's stored result, and return it, if
+// it contains the result of that operation. This would result in O(n).
```
Comment (naripok) [inline src/engine/filters/math.ts:8]: Isn't hash-table lookup O(1)?

File: src/engine/index.ts [hunk 3]
```diff
@@ -11,13 +11,31 @@ const check = (
   pageViews: PageView[],
   any = false
 ): boolean => {
-  const checkedConditions = conditions.map((condition) =>
-    evaluateCondition(condition, pageViews)
-  );
+  // Instead of going through the entire array again with an `.includes`
+  // method call, let's check _during_ other operations. This has a
+  // potential savings of `n` | n => conditions.length.
```
Comment (naripok) [inline src/engine/index.ts:16]: I agree here, but I would recommend the same using `some`/`every` for a cleaner implementation.

```
const check = (
  conditions: EngineCondition<AudienceDefinitionFilter>[],
  pageViews: PageView[],
  any = false
): boolean => {
  return conditions[any ? 'some' : 'every']((condition) =>
    evaluateCondition(condition, pageViews)
  );
};
```

File: src/engine/evaluate.ts (hunk 1 above)
Comment (CWelshE) [reply src/engine/evaluate.ts:32]: Oh, I hadn't realized that - I'll look into it and maybe make a performance test between the two. If that's the spec, then we might not get gains here. :thinking:

File: src/engine/index.ts (hunk 3 above)
Comment (CWelshE) [reply src/engine/index.ts:16]: Cool - sounds good to me. I'll refactor as such. Thanks!

**Pull request closed without merging**